File: sa_library/daxwriter.py
#!/usr/bin/python
#
# Writes dax files. Designed for use by analysis & simulation
# not so much for use by data taking code.
#
# Hazen 4/09
#

import numpy
import os

class DaxWriter():

    def __init__(self, name, w, h):
        self.name = name
        if len(os.path.dirname(name)) > 0:
            self.root_name = os.path.dirname(name) + "/" + os.path.splitext(os.path.basename(name))[0]
        else:
            self.root_name = os.path.splitext(os.path.basename(name))[0]
        self.w = int(w)
        self.h = int(h)
        self.fp = open(self.name, "wb")
        self.l = 0

    def addFrame(self, frame):
        frame = frame.copy()
        mask = (frame < 0)
        frame[mask] = 0
        mask = (frame > 65535)
        frame[mask] = 65535
        image16 = frame.astype('Int16')
        image16 = numpy.transpose(image16)
        image16.byteswap(True)
        image16.tofile(self.fp)
        self.l += 1
        if (self.w == 0) or (self.h == 0):
            [self.w, self.h] = frame.shape

    def close(self):
        self.fp.close()
        inf_fp = open(self.root_name + ".inf", "w")
        inf_fp.write("binning = 1 x 1\n")
        inf_fp.write("data type = 16 bit integers (binary, big endian)\n")
        inf_fp.write("frame dimensions = " + str(self.w) + " x " + str(self.h) + "\n")
        inf_fp.write("number of frames = " + str(self.l) + "\n")
        inf_fp.write("Lock Target = 0.0\n")
        if 1:
            inf_fp.write("x_start = 1\n")
            inf_fp.write("x_end = " + str(self.w) + "\n")
            inf_fp.write("y_start = 1\n")
            inf_fp.write("y_end = " + str(self.h) + "\n")
        inf_fp.close()

def dummyDaxFile(name, x_size, y_size):
    ddax = DaxWriter(name, x_size, y_size)
    frame = numpy.ones((x_size, y_size))
    ddax.addFrame(frame)
    ddax.close()

def singleFrameDax(name, frame):
    [fx, fy] = frame.shape
    dax_file = DaxWriter(name, fy, fx)
    dax_file.addFrame(frame)
    dax_file.close()

File: imagej_plugins/dax_file_io/Dax_Writer.java
//ImageJ writer plugin for Zhuang lab .dax files
//Translated from the Python versions in storm-analysis/sa_library and 
//based on sample IJ reader/writer plugins by Albert Cardona at 
//http://albert.rierol.net/imagej_programming_tutorials.html.
//Evan Heller, 10/2015

import java.io.*;  
import ij.*;  
import ij.io.*;
import ij.plugin.PlugIn;
import ij.process.ImageProcessor;
import ij.process.ImageConverter;

public class Dax_Writer implements PlugIn {  

    public void run(String arg) {  
        ImagePlus imp = WindowManager.getCurrentImage();  
        if (null == imp) return;  
        SaveDialog sd = new SaveDialog("Save Dax", "untitled", null);  
        String dir = sd.getDirectory();  
        if (null == dir) return; // user canceled dialog  
        dir = dir.replace('\\', '/'); // Windows safe  
        if (!dir.endsWith("/")) dir += "/";  
        saveDax(imp, dir + sd.getFileName());  
    }  

    static public void saveDax(ImagePlus imp, String path) {  
        File file = new File(path);  
        DataOutputStream dos = null;  

        try {  
            dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));  

            // read data:  
            FileInfo fi = imp.getFileInfo(); 
            String inf_file = path.split("\\.")[0] + ".inf";

            FileWriter fw = new FileWriter(inf_file); 

            //Write a rudimentary .inf file 
            fw.write("binning = 1 x 1\n");
            fw.write("data type = 16 bit integers (binary, big endian)\n");
            fw.write("frame dimensions = " + fi.width + " x " + fi.height + "\n");
            fw.write("number of frames = " + fi.nImages + "\n");
            fw.write("Lock Target = 0.0\n");
            fw.write("x_start = 1\n");
            fw.write("x_end = " + fi.width + "\n");
            fw.write("y_start = 1\n");
            fw.write("y_end = " + fi.height + "\n");
            fw.flush();
            fw.close();

            ImageProcessor ip;

            //Covert to 16-bit grayscale for writing to .dax
            if (imp.getBitDepth() != 16) {
                ImageConverter ic = new ImageConverter(imp);
                ic.convertToGray16();
            }

            //Write the data to file
            for (int i=0; i<imp.getNSlices(); i++) {  
                imp.setSlice(i);
                ip= imp.getProcessor(); //Get ip of current slice
                short[] px = (short[])ip.getPixels();

                for (int j=0; j<px.length; j++) 
                    dos.writeShort(px[j]);

            }  
            dos.flush();  
            dos.close(); 

        } catch (Exception e) {  
            e.printStackTrace();  
        } 
    }  
}  
